packages/babel-plugin-transform-regenerator/src/regenerator/emit.ts

```diff
@@ -56,1288 +19,1356 @@ const Ep = Emitter.prototype;
 // hitting memory limit of the machine it's running on. For comparison, the
 // estimated number of atoms in the observable universe is around 1e+80.
 const PENDING_LOCATION = Number.MAX_VALUE;
-Ep.loc = function () {
-  const l = util.getTypes().numericLiteral(PENDING_LOCATION);
-  this.insertedLocs.add(l);
-  return l;
-};
 
-Ep.getInsertedLocs = function () {
-  return this.insertedLocs;
-};
+function getDeclError(node: t.Node) {
+  return new Error(
+    "all declarations should have been transformed into " +
+      "assignments before the Exploder began its work: " +
+      JSON.stringify(node),
+  );
+}
 
-Ep.getContextId = function () {
-  return util.getTypes().clone(this.contextId);
+const catchParamVisitor: Visitor = {
+  Identifier: function (path, state: any) {
+    if (path.node.name === state.catchParamName && util.isReference(path)) {
+      util.replaceWithOrRemove(path, state.getSafeParam());
+    }
+  },
+
+  Scope: function (path, state: any) {
+    if (path.scope.hasOwnBinding(state.catchParamName)) {
+      // Don't descend into nested scopes that shadow the catch
+      // parameter with their own declarations.
+      path.skip();
+    }
+  },
 };
 
-// Sets the exact value of the given location to the offset of the next
-// Statement emitted.
-Ep.mark = function (loc: any) {
-  util.getTypes().assertLiteral(loc);
-  const index = this.listing.length;
-  if (loc.value === PENDING_LOCATION) {
-    loc.value = index;
-  } else {
-    // Locations can be marked redundantly, but their values cannot change
-    // once set the first time.
-    assert.strictEqual(loc.value, index);
+function isValidCompletion(record: any) {
```

Comment:
Let's define a `CompletionRecord` type, and just use TS to make sure we only build valid ones?